inference-engine/tests/functional/inference_engine/ov_async_infer_queue.cpp

```diff
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <gtest/gtest.h>
+#include <gmock/gmock-matchers.h>
+
+#include <map>
+
+#include <openvino/core/function.hpp>
+#include <openvino/core/shape.hpp>
+#include <openvino/core/type/element_type.hpp>
+
+#include <openvino/op/add.hpp>
+#include <openvino/op/multiply.hpp>
+#include <openvino/op/parameter.hpp>
+
+#include <openvino/runtime/core.hpp>
+#include <openvino/runtime/infer_request.hpp>
+#include <openvino/runtime/executable_network.hpp>
+#include <openvino/runtime/tensor.hpp>
+#include <openvino/runtime/async_infer_queue.hpp>
+
+using namespace ::testing;
+using namespace std;
+
+std::shared_ptr<ov::Function> create_add_model(ov::Shape shape) {
+    auto arg0 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto arg1 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto add = make_shared<ov::op::v1::Add>(arg0, arg1);
+
+    return make_shared<ov::Function>(add, ov::ParameterVector{arg0, arg1});
+}
+
+std::shared_ptr<ov::Function> create_mul_model(ov::Shape shape) {
+    auto arg0 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto arg1 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto mul = make_shared<ov::op::v1::Multiply>(arg0, arg1);
+
+    return make_shared<ov::Function>(mul, ov::ParameterVector{arg0, arg1});
+}
+
+TEST(AsyncInferQueueOVTests, BasicFlowTest) {
+    ov::runtime::Core core;
+    ov::runtime::ExecutableNetwork net =
+        core.compile_model(create_add_model({8}), "CPU");
+
+    ov::runtime::AsyncInferQueue basic_queue(net);
+
+    size_t counter = 0;
+    std::vector<ov::Any> callback_data;
+
+    auto f = [&counter, &callback_data](std::exception_ptr e,
+                                        ov::runtime::InferRequest &request,
+                                        const ov::Any &data) {
+        counter += 1;
+        callback_data.push_back(data);
+    };
+
+    basic_queue.set_callback(f);
+
+    size_t num_of_runs = 4;
+
+    for (size_t i = 0; i < num_of_runs; i++) {
+        basic_queue.start_async(i);
+    }
+
+    basic_queue.wait_all();
+
+    std::vector<size_t> converted_data;
+    std::transform(callback_data.begin(), callback_data.end(),
+                   std::back_inserter(converted_data),
+                   [](ov::Any val) -> std::size_t { return val.as<size_t>(); });
+
+    ASSERT_EQ(counter, num_of_runs);
+    ASSERT_THAT(converted_data, UnorderedElementsAreArray({0, 1, 2, 3}));
+}
+
+TEST(AsyncInferQueueOVTests, AccessRequestInsideCallbackTest) {
+    size_t num_of_runs = 5;
+    ov::Shape input_shape{1};
+    std::vector<float> results;
+    std::vector<float> expected_results;
+
+    ov::runtime::Core core;
+    ov::runtime::ExecutableNetwork net =
+        core.compile_model(create_add_model(input_shape), "CPU");
+
+    ov::runtime::AsyncInferQueue basic_queue(net, 3);
+
+    auto f = [&results](std::exception_ptr e,
```

Comment:
Do not omit error handling!

Reply:
Yep, I forgot about it as in Python errors where propagated and thrown almost seamlessly. Will do.